src/treeffuser/samples.py

```diff
@@ -64,55 +68,75 @@ def sample_kde(
         for i in tqdm(
             range(self.batch), disable=not verbose, desc="Fitting kernel densities for each x"
         ):
-            if self.ndim == 2:  # samples may have shape (n_samples, batch) when y_dim=1
-                y_i = self[:, i]
+            if self.ndim == 2:
+                y_i = self.samples[:, i, None]
             else:
-                y_i = self[:, i, :]
+                y_i = self.samples[:, i, :]
             kde = KernelDensity(bandwidth=bandwidth, algorithm="auto", kernel="gaussian")
             kde.fit(y_i)
             kdes.append(kde)
 
         return kdes
 
     def sample_max(self) -> Float[np.ndarray, "batch y_dim"]:
-        return super().max(axis=0)
-
-    def sample_mean(self) -> Float[np.ndarray, "batch y_dim"]:
-        return super().mean(axis=0)
+        return self.samples.max(axis=0)
 
     def sample_median(self) -> Float[np.ndarray, "batch y_dim"]:
-        return np.median(self, axis=0)
+        return np.median(self.samples, axis=0)
 
     def sample_min(self) -> Float[np.ndarray, "batch y_dim"]:
-        return super().min(axis=0)
+        return self.samples.min(axis=0)
 
     def sample_mode(
         self,
         bandwidth: Union[float, Literal["scott", "silverman"]] = 1.0,
         verbose: bool = False,
-    ) -> List[np.ndarray]:
-        _check_unidimensional(self)
+    ) -> Float[np.ndarray, "batch"]:
+        _check_unidimensional(self.samples)
 
         kdes = self.sample_kde(bandwidth=bandwidth)
 
         modes = []
         n_grid = np.max([2 * self.batch, 1000])  # heuristic for the grid granularity
         for i in tqdm(range(self.batch), disable=not verbose, desc="Searching for modes"):
-            if self.ndim == 2:
-                y_i = self[:, i]
+            if self.samples.ndim == 2:
+                y_i = self.samples[:, i, None]
             else:
-                y_i = self[:, i, :]
+                y_i = self.samples[:, i, :]
             grid = np.linspace(np.min(y_i), np.max(y_i), n_grid)
             log_density = kdes[i].score_samples(grid.reshape(-1, 1))
             modes.append(grid[np.argmax(log_density)])
+
+        modes = np.array(modes)
         return modes
 
     def sample_quantile(self, q) -> Float[np.ndarray, "q_dim batch y_dim"]:
-        return np.quantile(self, q, axis=0)
+        return np.quantile(self.samples, q, axis=0)
 
     def sample_range(self) -> Float[np.ndarray, "batch 2"]:
-        _check_unidimensional(self)
-        return np.stack((self.min(axis=0), self.max(axis=0)), axis=-1)
-
-    def sample_std(self, axis: int = 0) -> Float[np.ndarray, "batch y_dim"]:
-        return super().std(axis=0)
+        _check_unidimensional(self.samples)
+        return np.stack((self.samples.min(axis=0), self.samples.max(axis=0)), axis=-1)
+
+    def sample_std(self) -> Float[np.ndarray, "batch y_dim"]:
+        return self.samples.std(axis=0)
+
```

Comment:
```suggestion

    def to_numpy(self) -> Float[np.ndarray, "n_samples batch y_dim"]:
        return self.samples

```

Reply:
I don't think I agree with this change I think there should be just one way of doing it. Either this or accessing the attribute directly.

Reply:
Then I'd propose we use this and rename the samples attribute as `Samples._samples`. I believe `to_numpy()` is very readable and it aligns with standard practice (e.g. `DataFrame.to_numpy()` and `Tensor.numpy()`).